src/v/cluster/cluster_epoch_service.cc

```diff
@@ -0,0 +1,367 @@
+            vlog(clusterlog.warn, "Failed to force increment epoch: {}", ex);
+            continue;
+        }
+        auto result = result_fut.get();
+        if (result) {
+            _current_epoch = result.value().last_offset;
+            vlog(
+              clusterlog.debug,
+              "Forced increment of cluster epoch to: {}",
+              _current_epoch);
+            co_return;
+        } else if (result.error() != raft::errc::not_leader) {
+            vlog(
+              clusterlog.warn,
+              "Failed to force increment epoch: {}",
+              result.error());
+        }
+    }
+}
+
+cluster_epoch_service::cluster_epoch_service(
+  model::node_id self,
+  ss::sharded<rpc::connection_cache>* cc,
+  ss::sharded<partition_leaders_table>* l) noexcept
+  : _self(self)
+  , _rpc_conn(cc)
+  , _leaders(l) {}
+
+cluster_epoch_service::~cluster_epoch_service() noexcept = default;
+
+ss::future<> cluster_epoch_service::start() { co_return; }
+ss::future<> cluster_epoch_service::stop() {
+    co_await _gate.close();
+    if (!_shard0_state) {
+        co_return;
+    }
+    co_await _shard0_state->stop();
+    _shard0_state = nullptr;
+}
+
+void cluster_epoch_service::invalidate_epoch_cache(int64_t observed_epoch) {
+    _gate.check();
+    // This is safe to check without the lock, because we only use the lock to
+    // limit the number of cross shard calls and RPCs.
+    if (_cached_epoch == observed_epoch) {
+        _cached_epoch_time = ss::lowres_clock::time_point::min();
+        _epoch_updated_time = ss::lowres_clock::time_point::min();
+    }
+}
+
+ss::future<int64_t> cluster_epoch_service::get_cached_epoch() {
+    auto holder = _gate.hold();
+    // If the cache entry is needing an update, then block until
+    // that update is complete.
+    while (!_gate.is_closed() && cache_entry_needs_updated()) {
+        auto units = co_await _mu.get_units();
+        // Now that we have the lock, double check another fiber else did not
+        // update the cache while we were waiting for the lock.
+        if (cache_entry_needs_updated()) {
+            co_await update_epoch();
+        }
+    }
+    // If the cache entry is expired, then we can spawn an update in the
+    // background to update the epoch. Use the lock to ensure we only have
+    // a single
+    if (cache_entry_expired()) {
+        auto maybe_units = _mu.try_get_units();
+        if (maybe_units) {
+            ssx::spawn_with_gate(
+              _gate, [this, units = std::move(*maybe_units)]() mutable {
+                  return update_epoch().finally([units = std::move(units)] {});
+              });
+        }
+    }
+    co_return _cached_epoch;
+}
+
+ss::future<> cluster_epoch_service::update_epoch() {
+    auto maybe_epoch = co_await container().invoke_on(
+      controller_stm_shard, &cluster_epoch_service::get_current_epoch);
+    if (!maybe_epoch) {
+        try {
+            maybe_epoch = co_await fetch_leader_epoch();
```

Comment:
wondering if every shard should ddos the leader..  since all the shards are routing to `controller_stm_shard`, it may just fetch once from the leader (if it is not the leader)?

Reply:
I think there are two approaches:

1. change the fetches on other shards to ask shard0 for it's epoch. Maybe cache invalidation becomes tricky.

Reply:
Went with approach #1 for now.